### PR TITLE
[extractor/Douyin] fix Douyin

### PR DESCRIPTION
**Pull request by 114514ns**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

fix Douyin exactor 

Fixes #
https://github.com/yt-dlp/yt-dlp/issues/8139
https://github.com/yt-dlp/yt-dlp/issues/8090
https://github.com/yt-dlp/yt-dlp/issues/7854

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

**Comment by TheAznCoderPro** [comment]
How do you pull-request a pull-request? (to fix the linter warnings)

**Comment by bashonly** [comment]
I would advise not doing anything with this code until the author finishes the PR checklist

**Comment by TheAznCoderPro** [comment]
@114514ns make sure to complete the template (the box about the licences)

**Comment by yifeikong** [comment]
Any updates?

**Comment by TheAznCoderPro** [comment]
> Any updates?

PR seems like it should be good to merge

File: yt_dlp/extractor/tiktok.py [hunk 1]
```diff
@@ -1035,24 +1060,17 @@ class DouyinIE(TikTokBaseIE):
...
-            e.expected = True
-            self.to_screen(f'{e}; trying with webpage')
-
-        webpage = self._download_webpage(url, video_id)
-        render_data = self._search_json(
-            r'<script [^>]*\bid=[\'"]RENDER_DATA[\'"][^>]*>', webpage, 'render data', video_id,
-            contains_pattern=r'%7B(?s:.+)%7D', fatal=False, transform_source=compat_urllib_parse_unquote)
+        webpage = self._download_webpage("https://www.douyin.com/aweme/v1/web/aweme/detail/?aweme_id=" + video_id,
+                                         video_id)
+        render_data = self._parse_json(webpage, video_id, fatal=False)
```
Comment (bashonly) [inline yt_dlp/extractor/tiktok.py:1065]:
```suggestion
        detail = traverse_obj(self._download_json(
            'https://www.douyin.com/aweme/v1/web/aweme/detail/', video_id,
            'Downloading web detail JSON', 'Failed to download web detail JSON',
            query={'aweme_id': video_id}, fatal=False), ('aweme_detail', {dict}))
```

File: yt_dlp/extractor/tiktok.py [hunk 2]
```diff
@@ -1035,24 +1060,17 @@ class DouyinIE(TikTokBaseIE):
...
         if not render_data:
             # TODO: Run verification challenge code to generate signature cookies
             cookies = self._get_cookies(self._WEBPAGE_HOST)
             expected = not cookies.get('s_v_web_id') or not cookies.get('ttwid')
             raise ExtractorError(
                 'Fresh cookies (not necessarily logged in) are needed', expected=expected)
```
Comment (bashonly) [inline yt_dlp/extractor/tiktok.py:1071]:
```suggestion
        if not detail:
            # TODO: Run verification challenge code to generate signature cookies
            raise ExtractorError(
                'Fresh cookies (not necessarily logged in) are needed',
                expected=not self._get_cookies(self._WEBPAGE_HOST).get('s_v_web_id'))
```

File: yt_dlp/extractor/tiktok.py [hunk 3]
```diff
@@ -1035,24 +1060,17 @@ class DouyinIE(TikTokBaseIE):
...
 
-        return self._parse_aweme_video_web(get_first(render_data, ('aweme', 'detail')), url, video_id)
+        return self._parse_aweme_video_web(traverse_obj(render_data, 'aweme_detail'), url, video_id)
```
Comment (bashonly) [inline yt_dlp/extractor/tiktok.py:1073]: wrong method
```suggestion
        return self._parse_aweme_video_app(detail)
```

File: yt_dlp/extractor/tiktok.py [hunk 4]
```diff
@@ -767,7 +790,8 @@ def _video_entries_api(self, webpage, user_id, username):
             'max_cursor': 0,
             'min_cursor': 0,
             'retry_type': 'no_retry',
-            'device_id': ''.join(random.choices(string.digits, k=19)),  # Some endpoints don't like randomized device_id, so it isn't directly set in _call_api.
+            'device_id': ''.join(random.choices(string.digits, k=19)),
+            # Some endpoints don't like randomized device_id, so it isn't directly set in _call_api.
```
Comment (bashonly) [inline yt_dlp/extractor/tiktok.py:794]: revert unnecessary/unrelated formatting change
```suggestion
            'device_id': ''.join(random.choices(string.digits, k=19)), # Some endpoints don't like randomized device_id, so it isn't directly set in _call_api.
```

File: yt_dlp/extractor/tiktok.py [hunk 5]
```diff
@@ -6,16 +6,15 @@
 import time
 
 from .common import InfoExtractor
-from ..compat import compat_urllib_parse_unquote, compat_urllib_parse_urlparse
 from ..networking import HEADRequest
+from ..compat import compat_urllib_parse_urlparse
```
Comment (bashonly) [inline yt_dlp/extractor/tiktok.py:10]: revert unnecessary/unrelated formatting change
```suggestion
from ..compat import compat_urllib_parse_urlparse
from ..networking import HEADRequest
```

File: yt_dlp/extractor/tiktok.py [hunk 6]
```diff
@@ -112,7 +111,8 @@ def _call_api(self, ep, query, video_id, fatal=True,
                 self._WORKING_APP_VERSION = (app_version, manifest_app_version)
                 self.write_debug('Imported app version combo from extractor arguments')
             elif app_version or manifest_app_version:
-                self.report_warning('Only one of the two required version params are passed as extractor arguments', only_once=True)
+                self.report_warning('Only one of the two required version params are passed as extractor arguments',
+                                    only_once=True)
```
Comment (bashonly) [inline yt_dlp/extractor/tiktok.py:115]: revert unnecessary/unrelated formatting change
```suggestion
                self.report_warning('Only one of the two required version params are passed as extractor arguments', only_once=True)
```

File: yt_dlp/extractor/tiktok.py [hunk 7]
```diff
@@ -133,7 +133,8 @@ def _call_api(self, ep, query, video_id, fatal=True,
                         else:
                             self.report_warning(str(e.cause or e.msg))
                             return
-                    self.report_warning('%s. Retrying... (attempt %s of %s)' % (str(e.cause or e.msg), count, len(self._APP_VERSIONS)))
+                    self.report_warning(
+                        '%s. Retrying... (attempt %s of %s)' % (str(e.cause or e.msg), count, len(self._APP_VERSIONS)))
```
Comment (bashonly) [inline yt_dlp/extractor/tiktok.py:137]: revert unnecessary/unrelated formatting change
```suggestion
                    self.report_warning('%s. Retrying... (attempt %s of %s)' % (str(e.cause or e.msg), count, len(self._APP_VERSIONS)))
```

File: yt_dlp/extractor/tiktok.py [hunk 8]
```diff
@@ -151,7 +152,8 @@ def _get_subtitles(self, aweme_detail, aweme_id):
         subtitles = {}
         # aweme/detail endpoint subs
         captions_info = traverse_obj(
-            aweme_detail, ('interaction_stickers', ..., 'auto_video_caption_info', 'auto_captions', ...), expected_type=dict)
+            aweme_detail, ('interaction_stickers', ..., 'auto_video_caption_info', 'auto_captions', ...),
+            expected_type=dict)
```
Comment (bashonly) [inline yt_dlp/extractor/tiktok.py:156]: revert unnecessary/unrelated formatting change
```suggestion
            aweme_detail, ('interaction_stickers', ..., 'auto_video_caption_info', 'auto_captions', ...), expected_type=dict)
```

File: yt_dlp/extractor/tiktok.py [hunk 9]
```diff
@@ -242,7 +244,8 @@ def extract_addr(addr, add_meta={}):
                 'format_id': 'play_addr',
                 'format_note': 'Direct video',
                 'vcodec': 'h265' if traverse_obj(
-                    video_info, 'is_bytevc1', 'is_h265') else 'h264',  # TODO: Check for "direct iOS" videos, like https://www.tiktok.com/@cookierun_dev/video/7039716639834656002
+                    video_info, 'is_bytevc1', 'is_h265') else 'h264',
+                # TODO: Check for "direct iOS" videos, like https://www.tiktok.com/@cookierun_dev/video/7039716639834656002
```
Comment (bashonly) [inline yt_dlp/extractor/tiktok.py:248]: revert unnecessary/unrelated formatting change
```suggestion
                    video_info, 'is_bytevc1', 'is_h265') else 'h264', # TODO: Check for "direct iOS" videos, like https://www.tiktok.com/@cookierun_dev/video/7039716639834656002
```

File: yt_dlp/extractor/tiktok.py [hunk 10]
```diff
@@ -307,7 +310,8 @@ def extract_addr(addr, add_meta={}):
         contained_music_author = traverse_obj(
             music_info, ('matched_song', 'author'), ('matched_pgc_sound', 'author'), 'author', expected_type=str)
 
-        is_generic_og_trackname = music_info.get('is_original_sound') and music_info.get('title') == 'original sound - %s' % music_info.get('owner_handle')
+        is_generic_og_trackname = music_info.get('is_original_sound') and music_info.get(
+            'title') == 'original sound - %s' % music_info.get('owner_handle')
```
Comment (bashonly) [inline yt_dlp/extractor/tiktok.py:314]: revert unnecessary/unrelated formatting change
```suggestion
        is_generic_og_trackname = music_info.get('is_original_sound') and music_info.get('title') == 'original sound - %s' % music_info.get('owner_handle')
```

File: yt_dlp/extractor/tiktok.py [hunk 11]
```diff
@@ -362,22 +366,41 @@ def _parse_aweme_video_web(self, aweme_detail, webpage_url, video_id):
...
-                'height': height,
-            })
+        if ("tiktok" in webpage_url):
+            for play_url in traverse_obj(video_info, ('playAddr', ((..., 'src'), None), {url_or_none})):
+                formats.append({
+                    'url': self._proto_relative_url(play_url),
+                    'ext': 'mp4',
+                    'width': width,
+                    'height': height,
+                })
```
Comment (bashonly) [inline yt_dlp/extractor/tiktok.py:376]: revert
```suggestion
        for play_url in traverse_obj(video_info, ('playAddr', ((..., 'src'), None), {url_or_none})):
            formats.append({
                'url': self._proto_relative_url(play_url),
                'ext': 'mp4',
                'width': width,
                'height': height,
            })
```

File: yt_dlp/extractor/tiktok.py [hunk 12]
```diff
@@ -362,22 +366,41 @@ def _parse_aweme_video_web(self, aweme_detail, webpage_url, video_id):
...
+
+            for download_url in traverse_obj(video_info, ('download_addr', 'url_list', ...)):
+                formats.append({
+                    'format_id': 'download',
+                    'url': download_url,
+                    'ext': 'mp4',
+                    'width': width,
+                    'height': height,
+                })
```
Comment (bashonly) [inline yt_dlp/extractor/tiktok.py:403]: revert
```suggestion
        for download_url in traverse_obj(video_info, (('downloadAddr', ('download', 'url')), {url_or_none})):
            formats.append({
                'format_id': 'download',
                'url': self._proto_relative_url(download_url),
                'ext': 'mp4',
                'width': width,
                'height': height,
            })
```

**Review by bashonly — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/tiktok.py [hunk 13]
```diff
@@ -242,7 +241,7 @@ def extract_addr(addr, add_meta={}):
                 'format_id': 'play_addr',
                 'format_note': 'Direct video',
                 'vcodec': 'h265' if traverse_obj(
-                    video_info, 'is_bytevc1', 'is_h265') else 'h264',  # TODO: Check for "direct iOS" videos, like https://www.tiktok.com/@cookierun_dev/video/7039716639834656002
+                    video_info, 'is_bytevc1', 'is_h265') else 'h264', # TODO: Check for "direct iOS" videos, like https://www.tiktok.com/@cookierun_dev/video/7039716639834656002
```
Comment (bashonly) [inline yt_dlp/extractor/tiktok.py:244]:
```suggestion
                    video_info, 'is_bytevc1', 'is_h265') else 'h264',  # TODO: Check for "direct iOS" videos, like https://www.tiktok.com/@cookierun_dev/video/7039716639834656002
```

File: yt_dlp/extractor/tiktok.py [hunk 14]
```diff
@@ -767,7 +766,7 @@ def _video_entries_api(self, webpage, user_id, username):
             'max_cursor': 0,
             'min_cursor': 0,
             'retry_type': 'no_retry',
-            'device_id': ''.join(random.choices(string.digits, k=19)),  # Some endpoints don't like randomized device_id, so it isn't directly set in _call_api.
+            'device_id': ''.join(random.choices(string.digits, k=19)), # Some endpoints don't like randomized device_id, so it isn't directly set in _call_api.
```
Comment (bashonly) [inline yt_dlp/extractor/tiktok.py:769]:
```suggestion
            'device_id': ''.join(random.choices(string.digits, k=19)),  # Some endpoints don't like randomized device_id, so it isn't directly set in _call_api.
```

**Review by bashonly — CHANGES_REQUESTED** [review]

**Review by bashonly — APPROVED** [review]

File: yt_dlp/extractor/tiktok.py [hunk 15]
```diff
@@ -237,22 +243,27 @@ def extract_addr(addr, add_meta={}):
...
-                'width': video_info.get('width'),
-                'height': video_info.get('height'),
+                'width': width,
+                'height': height,
             }))
         if video_info.get('download_addr'):
-            formats.extend(extract_addr(video_info['download_addr'], {
+            download_addr = video_info['download_addr']
+            download_width = int_or_none(download_addr.get('width'))
+            download_height = self.RESOLUTION_MAP.get(download_width)
```
Comment (bashonly) [inline yt_dlp/extractor/tiktok.py:260]:
```suggestion
            download_height = self.RESOLUTION_MAP.get(download_width)  # download_addr['height'] is inaccurate
```

**Pull request merged**